ratelimits/source-redis_test.go

```diff
@@ -0,0 +1,342 @@
+package ratelimits
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/metrics"
+	"github.com/letsencrypt/boulder/test"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/jmhodges/clock"
+)
+
+const (
+	tenZeroZeroThree = "10.0.0.3"
+	tenZeroZeroFour  = "10.0.0.4"
+	tenZeroZeroFive  = "10.0.0.5"
+	tenZeroZeroSix   = "10.0.0.6"
+	tenZeroZeroSeven = "10.0.0.7"
+)
+
+func makeClient() (*RedisSource, clock.FakeClock) {
+	CACertFile := "../test/redis-tls/minica.pem"
+	CertFile := "../test/redis-tls/boulder/cert.pem"
+	KeyFile := "../test/redis-tls/boulder/key.pem"
+	tlsConfig := cmd.TLSConfig{
+		CACertFile: CACertFile,
+		CertFile:   CertFile,
+		KeyFile:    KeyFile,
+	}
+	tlsConfig2, err := tlsConfig.Load(metrics.NoopRegisterer)
+	if err != nil {
+		panic(err)
+	}
+
+	client := redis.NewRing(&redis.RingOptions{
+		Addrs: map[string]string{
+			"shard1": "10.33.33.4:4218",
+			"shard2": "10.33.33.5:4218",
+		},
+		Username:  "unittest-rw",
+		Password:  "824968fa490f4ecec1e52d5e34916bdb60d45f8d",
+		TLSConfig: tlsConfig2,
+	})
+	clk := clock.NewFake()
+	return NewRedisSource(client, 5*time.Second, clk, metrics.NoopRegisterer), clk
+}
+
+// newTestLimiter makes a new limiter with the following configuration:
+//   - 'NewRegistrationsPerIPAddress' burst: 20 count: 20 period: 1s
+func newRedisTestLimiter(t *testing.T) (*Limiter, clock.FakeClock, func(Name, string)) {
+	source, clk := makeClient()
+	l, err := NewLimiter(clk, source, "testdata/working_default.yml", "")
+	test.AssertNotError(t, err, "should not error")
+	return l, clk, func(name Name, id string) {
+		err = l.Reset(context.Background(), name, id)
+		if err != nil {
+			t.Fatalf("failed to reset bucket: %v", err)
+		}
+	}
+}
+
+// newRedisTestLimiterWithOverrides makes a new limiter with the following
+// configuration:
+//   - 'NewRegistrationsPerIPAddress' burst: 20 count: 20 period: 1s
+//   - 'NewRegistrationsPerIPAddress:10.0.0.2' burst: 40 count: 40 period: 1s
+func newRedisTestLimiterWithOverrides(t *testing.T) (*Limiter, clock.FakeClock, func(Name, string)) {
+	source, clk := makeClient()
+	l, err := NewLimiter(clk, source, "testdata/working_default.yml", "testdata/working_override.yml")
+	test.AssertNotError(t, err, "should not error")
+	return l, clk, func(name Name, id string) {
+		err = l.Reset(context.Background(), name, id)
+		if err != nil {
+			t.Fatalf("failed to reset bucket: %v", err)
+		}
+	}
+}
+
+func Test_Redis_Limiter_Refund_and_Spend_cost_err(t *testing.T) {
+	t.Parallel()
```

Comment:
I think the only reason this test is parallelizable is because it puts a unique IP address (tenZeroZeroOne, while the other tests uses Two, etc) in the bucket keys it cares about. I think a better approach would be for each test to generate its own fully-random IP address to use, giving 1-in-2<sup>32</sup> chances of a collision but zero maintenance burden when tests are added or modified in the future.